File: backend/models/pricing.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// PriceRule describes a single pricing adjustment within a PriceSet.
type PriceRule struct {
	Kind        RuleKind  `json:"kind" bson:"kind"`
	Amount      float64   `json:"amount" bson:"amount"`
	Scope       RuleScope `json:"scope" bson:"scope"`
	ScopeRefs   []string  `json:"scopeRefs,omitempty" bson:"scopeRefs,omitempty"`
	MaxDiscount float64   `json:"maxDiscount,omitempty" bson:"maxDiscount,omitempty"`
	Priority    int       `json:"priority" bson:"priority"`
}

// RuleKind mirrors the standard pricing operators.
type RuleKind string

const (
	RuleKindPercentage RuleKind = "percentage"
	RuleKindAbsolute   RuleKind = "absolute"
	RuleKindOverride   RuleKind = "override"
)

// RuleScope controls which items a rule targets.
type RuleScope string

const (
	RuleScopeAll      RuleScope = "all"
	RuleScopeProduct  RuleScope = "product"
	RuleScopeCategory RuleScope = "category"
	RuleScopeVariant  RuleScope = "variant"
	RuleScopeSKU      RuleScope = "sku"
)

// PriceConditions gate whether a PriceSet applies to a given pricing context.
// All non-zero conditions must match (AND semantics).
type PriceConditions struct {
	CouponCode  string   `json:"couponCode,omitempty" bson:"couponCode,omitempty"`
	MinSubtotal float64  `json:"minSubtotal,omitempty" bson:"minSubtotal,omitempty"`
	MinQuantity int      `json:"minQuantity,omitempty" bson:"minQuantity,omitempty"`
	CustomerTier string `json:"customerTier,omitempty" bson:"customerTier,omitempty"`
	CustomerIDs []string `json:"customerIDs,omitempty" bson:"customerIDs,omitempty"`
}

// PriceSet is an ordered, condition-gated bundle of price rules.
type PriceSet struct {
	ID                primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	Name              string             `json:"name" bson:"name"`
	Description       string             `json:"description,omitempty" bson:"description,omitempty"`
	Priority          int                `json:"priority" bson:"priority"`
	StopFurtherRules  bool               `json:"stopFurtherRules" bson:"stopFurtherRules"`
	Conditions        PriceConditions    `json:"conditions" bson:"conditions"`
	Rules             []PriceRule        `json:"rules" bson:"rules"`
	MaxUses           int                `json:"maxUses,omitempty" bson:"maxUses,omitempty"`
	UsedCount         int                `json:"usedCount" bson:"usedCount"`
	MaxUsesPerCustomer int               `json:"maxUsesPerCustomer,omitempty" bson:"maxUsesPerCustomer,omitempty"`
	Active            bool               `json:"active" bson:"active"`
	StartsAt          *time.Time         `json:"startsAt,omitempty" bson:"startsAt,omitempty"`
	EndsAt            *time.Time         `json:"endsAt,omitempty" bson:"endsAt,omitempty"`
	CreatedAt         time.Time          `json:"createdAt" bson:"createdAt"`
	UpdatedAt         time.Time          `json:"updatedAt" bson:"updatedAt"`
}

// PriceSchedule applies a time-based price formula to a scope (inflation/market).
type PriceSchedule struct {
	ID            primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	Name          string             `json:"name" bson:"name"`
	Description   string             `json:"description,omitempty" bson:"description,omitempty"`
	Scope         ScheduleScope      `json:"scope" bson:"scope"`
	ScopeRefs     []string           `json:"scopeRefs,omitempty" bson:"scopeRefs,omitempty"`
	Mode          ScheduleMode       `json:"mode" bson:"mode"`
	Value         float64            `json:"value" bson:"value"`
	Priority      int                `json:"priority" bson:"priority"`
	Active        bool               `json:"active" bson:"active"`
	EffectiveFrom time.Time          `json:"effectiveFrom" bson:"effectiveFrom"`
	EffectiveTo   *time.Time         `json:"effectiveTo,omitempty" bson:"effectiveTo,omitempty"`
	CreatedAt     time.Time          `json:"createdAt" bson:"createdAt"`
	UpdatedAt     time.Time          `json:"updatedAt" bson:"updatedAt"`
}

type ScheduleScope string

const (
	ScheduleScopeGlobal   ScheduleScope = "global"
	ScheduleScopeCategory ScheduleScope = "category"
	ScheduleScopeProduct  ScheduleScope = "product"
	ScheduleScopeVariant  ScheduleScope = "variant"
)

type ScheduleMode string

const (
	ScheduleModePercentage ScheduleMode = "percentage"
	ScheduleModeAbsolute   ScheduleMode = "absolute"
	ScheduleModeFixed      ScheduleMode = "fixed"
)

// PriceHistoryEntry records a single effective price change (audit trail).
type PriceHistoryEntry struct {
	ProductID   string    `json:"productId" bson:"productId"`
	VariantID   string    `json:"variantId,omitempty" bson:"variantId,omitempty"`
	OldPrice    float64   `json:"oldPrice" bson:"oldPrice"`
	NewPrice    float64   `json:"newPrice" bson:"newPrice"`
	Reason      string    `json:"reason,omitempty" bson:"reason,omitempty"`
	SourceType  string    `json:"sourceType" bson:"sourceType"`
	SourceRef   string    `json:"sourceRef,omitempty" bson:"sourceRef,omitempty"`
	EffectiveAt time.Time `json:"effectiveAt" bson:"effectiveAt"`
}

// AppliedPriceRule records which set/rule produced a discount (for order Pricing map).
type AppliedPriceRule struct {
	SetID    string   `json:"setId" bson:"setId"`
	SetName  string   `json:"setName" bson:"setName"`
	Kind     RuleKind `json:"kind" bson:"kind"`
	Amount   float64  `json:"amount" bson:"amount"`
	Discount float64  `json:"discount" bson:"discount"`
}

File: admin_console/lib/models/pricing.dart
// Admin pricing models mirroring backend/models/pricing.go JSON.

enum PriceRuleKind {
  percentage,
  absolute,
  override;

  static PriceRuleKind fromName(String? name) {
    switch (name) {
      case 'absolute':
        return PriceRuleKind.absolute;
      case 'override':
        return PriceRuleKind.override;
      default:
        return PriceRuleKind.percentage;
    }
  }

  String get name => switch (this) {
        PriceRuleKind.percentage => 'percentage',
        PriceRuleKind.absolute => 'absolute',
        PriceRuleKind.override => 'override',
      };
}

enum RuleScope {
  all,
  product,
  category,
  variant,
  sku;

  static RuleScope fromName(String? name) {
    switch (name) {
      case 'product':
        return RuleScope.product;
      case 'category':
        return RuleScope.category;
      case 'variant':
        return RuleScope.variant;
      case 'sku':
        return RuleScope.sku;
      default:
        return RuleScope.all;
    }
  }

  String get name => switch (this) {
        RuleScope.all => 'all',
        RuleScope.product => 'product',
        RuleScope.category => 'category',
        RuleScope.variant => 'variant',
        RuleScope.sku => 'sku',
      };
}

class PriceRule {
  final PriceRuleKind kind;
  final double amount;
  final RuleScope scope;
  final List<String> scopeRefs;
  final double maxDiscount;
  final int priority;

  PriceRule({
    required this.kind,
    this.amount = 0,
    this.scope = RuleScope.all,
    this.scopeRefs = const [],
    this.maxDiscount = 0,
    this.priority = 1,
  });

  factory PriceRule.fromJson(Map<String, dynamic> json) {
    return PriceRule(
      kind: PriceRuleKind.fromName(json['kind']?.toString()),
      amount: (json['amount'] as num?)?.toDouble() ?? 0,
      scope: RuleScope.fromName(json['scope']?.toString()),
      scopeRefs: (json['scopeRefs'] as List?)?.map((e) => e.toString()).toList() ?? const [],
      maxDiscount: (json['maxDiscount'] as num?)?.toDouble() ?? 0,
      priority: (json['priority'] as num?)?.toInt() ?? 1,
    );
  }

  Map<String, dynamic> toJson() => {
        'kind': kind.name,
        'amount': amount,
        'scope': scope.name,
        if (scopeRefs.isNotEmpty) 'scopeRefs': scopeRefs,
        if (maxDiscount > 0) 'maxDiscount': maxDiscount,
        'priority': priority,
      };
}

class PriceConditions {
  final String couponCode;
  final double minSubtotal;
  final int minQuantity;
  final String customerTier;
  final List<String> customerIds;

  const PriceConditions({
    this.couponCode = '',
    this.minSubtotal = 0,
    this.minQuantity = 0,
    this.customerTier = '',
    this.customerIds = const [],
  });

  factory PriceConditions.fromJson(Map<String, dynamic>? json) {
    if (json == null) return PriceConditions();
    return PriceConditions(
      couponCode: json['couponCode']?.toString() ?? '',
      minSubtotal: (json['minSubtotal'] as num?)?.toDouble() ?? 0,
      minQuantity: (json['minQuantity'] as num?)?.toInt() ?? 0,
      customerTier: json['customerTier']?.toString() ?? '',
      customerIds: (json['customerIDs'] as List?)?.map((e) => e.toString()).toList() ?? const [],
    );
  }

  Map<String, dynamic> toJson() => {
        if (couponCode.isNotEmpty) 'couponCode': couponCode,
        if (minSubtotal > 0) 'minSubtotal': minSubtotal,
        if (minQuantity > 0) 'minQuantity': minQuantity,
        if (customerTier.isNotEmpty) 'customerTier': customerTier,
        if (customerIds.isNotEmpty) 'customerIDs': customerIds,
      };
}

class PriceSet {
  final String id;
  final String name;
  final String description;
  final int priority;
  final bool stopFurtherRules;
  final PriceConditions conditions;
  final List<PriceRule> rules;
  final int maxUses;
  final int usedCount;
  final int maxUsesPerCustomer;
  final bool active;
  final DateTime? startsAt;
  final DateTime? endsAt;

  PriceSet({
    this.id = '',
    required this.name,
    this.description = '',
    this.priority = 1,
    this.stopFurtherRules = false,
    this.conditions = const PriceConditions(),
    this.rules = const [],
    this.maxUses = 0,
    this.usedCount = 0,
    this.maxUsesPerCustomer = 0,
    this.active = true,
    this.startsAt,
    this.endsAt,
  });

  factory PriceSet.fromJson(Map<String, dynamic> json) {
    return PriceSet(
      id: json['id']?.toString() ?? '',
      name: json['name']?.toString() ?? '',
      description: json['description']?.toString() ?? '',
      priority: (json['priority'] as num?)?.toInt() ?? 1,
      stopFurtherRules: json['stopFurtherRules'] == true,
      conditions: PriceConditions.fromJson(
          (json['conditions'] as Map<String, dynamic>?)),
      rules: (json['rules'] as List?)
              ?.map((e) => PriceRule.fromJson(e as Map<String, dynamic>))
              .toList() ??
          const [],
      maxUses: (json['maxUses'] as num?)?.toInt() ?? 0,
      usedCount: (json['usedCount'] as num?)?.toInt() ?? 0,
      maxUsesPerCustomer: (json['maxUsesPerCustomer'] as num?)?.toInt() ?? 0,
      active: json['active'] ?? true,
      startsAt: json['startsAt'] != null
          ? DateTime.tryParse(json['startsAt'].toString())
          : null,
      endsAt: json['endsAt'] != null
          ? DateTime.tryParse(json['endsAt'].toString())
          : null,
    );
  }

  Map<String, dynamic> toJson() => {
        'name': name,
        if (description.isNotEmpty) 'description': description,
        'priority': priority,
        'stopFurtherRules': stopFurtherRules,
        'conditions': conditions.toJson(),
        'rules': rules.map((r) => r.toJson()).toList(),
        if (maxUses > 0) 'maxUses': maxUses,
        if (maxUsesPerCustomer > 0) 'maxUsesPerCustomer': maxUsesPerCustomer,
        'active': active,
        if (startsAt != null) 'startsAt': startsAt!.toIso8601String(),
        if (endsAt != null) 'endsAt': endsAt!.toIso8601String(),
      };
}

enum ScheduleScope {
  global,
  category,
  product,
  variant;

  static ScheduleScope fromName(String? name) {
    switch (name) {
      case 'category':
        return ScheduleScope.category;
      case 'product':
        return ScheduleScope.product;
      case 'variant':
        return ScheduleScope.variant;
      default:
        return ScheduleScope.global;
    }
  }

  String get name => switch (this) {
        ScheduleScope.global => 'global',
        ScheduleScope.category => 'category',
        ScheduleScope.product => 'product',
        ScheduleScope.variant => 'variant',
      };
}

enum ScheduleMode {
  percentage,
  absolute,
  fixed;

  static ScheduleMode fromName(String? name) {
    switch (name) {
      case 'absolute':
        return ScheduleMode.absolute;
      case 'fixed':
        return ScheduleMode.fixed;
      default:
        return ScheduleMode.percentage;
    }
  }

  String get name => switch (this) {
        ScheduleMode.percentage => 'percentage',
        ScheduleMode.absolute => 'absolute',
        ScheduleMode.fixed => 'fixed',
      };
}

class PriceSchedule {
  final String id;
  final String name;
  final String description;
  final ScheduleScope scope;
  final List<String> scopeRefs;
  final ScheduleMode mode;
  final double value;
  final int priority;
  final bool active;
  final DateTime effectiveFrom;
  final DateTime? effectiveTo;

  PriceSchedule({
    this.id = '',
    required this.name,
    this.description = '',
    this.scope = ScheduleScope.global,
    this.scopeRefs = const [],
    this.mode = ScheduleMode.percentage,
    required this.value,
    this.priority = 1,
    this.active = true,
    required this.effectiveFrom,
    this.effectiveTo,
  });

  factory PriceSchedule.fromJson(Map<String, dynamic> json) {
    return PriceSchedule(
      id: json['id']?.toString() ?? '',
      name: json['name']?.toString() ?? '',
      description: json['description']?.toString() ?? '',
      scope: ScheduleScope.fromName(json['scope']?.toString()),
      scopeRefs: (json['scopeRefs'] as List?)?.map((e) => e.toString()).toList() ?? const [],
      mode: ScheduleMode.fromName(json['mode']?.toString()),
      value: (json['value'] as num?)?.toDouble() ?? 0,
      priority: (json['priority'] as num?)?.toInt() ?? 1,
      active: json['active'] ?? true,
      effectiveFrom: DateTime.tryParse(json['effectiveFrom']?.toString() ?? '') ?? DateTime.now(),
      effectiveTo: json['effectiveTo'] != null
          ? DateTime.tryParse(json['effectiveTo'].toString())
          : null,
    );
  }

  Map<String, dynamic> toJson() => {
        'name': name,
        if (description.isNotEmpty) 'description': description,
        'scope': scope.name,
        if (scopeRefs.isNotEmpty) 'scopeRefs': scopeRefs,
        'mode': mode.name,
        'value': value,
        'priority': priority,
        'active': active,
        'effectiveFrom': effectiveFrom.toIso8601String(),
        if (effectiveTo != null) 'effectiveTo': effectiveTo!.toIso8601String(),
      };
}

class PriceHistoryEntry {
  final String productId;
  final String variantId;
  final double oldPrice;
  final double newPrice;
  final String reason;
  final String sourceType;
  final String sourceRef;
  final DateTime effectiveAt;

  PriceHistoryEntry({
    required this.productId,
    this.variantId = '',
    required this.oldPrice,
    required this.newPrice,
    this.reason = '',
    this.sourceType = '',
    this.sourceRef = '',
    required this.effectiveAt,
  });

  factory PriceHistoryEntry.fromJson(Map<String, dynamic> json) {
    return PriceHistoryEntry(
      productId: json['productId']?.toString() ?? '',
      variantId: json['variantId']?.toString() ?? '',
      oldPrice: (json['oldPrice'] as num?)?.toDouble() ?? 0,
      newPrice: (json['newPrice'] as num?)?.toDouble() ?? 0,
      reason: json['reason']?.toString() ?? '',
      sourceType: json['sourceType']?.toString() ?? '',
      sourceRef: json['sourceRef']?.toString() ?? '',
      effectiveAt: DateTime.tryParse(json['effectiveAt']?.toString() ?? '') ?? DateTime.now(),
    );
  }
}